crates/egui/src/text_selection/text_cursor_state.rs

```diff
@@ -302,18 +302,42 @@ pub fn find_line_start(text: &str, current_index: CCursor) -> CCursor {
     // number of multi byte chars.
     // We need to know the char index to be able to correctly set the cursor
     // later.
-    let chars_count = text.chars().count();
+    let mut char_line_indexes: Vec<usize> = create_char_line_indexes(text);
+    let line_start_index = get_line_start_index(&mut char_line_indexes, current_index.index);
 
-    let position = text
-        .chars()
-        .rev()
-        .skip(chars_count - current_index.index)
-        .position(|x| x == '\n');
+    CCursor::new(line_start_index)
+}
+
+pub fn create_char_line_indexes(text: &str) -> Vec<usize> {
```

Comment:
If you use `CCursor` instead of `usize`, the code becomes self-documenting

```suggestion
pub fn create_char_line_indexes(text: &str) -> Vec<CCursor> {
```

Reply:
If CCursor is used instead of usize,                                                  
it seems handling functions might become difficult,                                   
so I'm not sure if it's a good approach.                                              
I will respect your final opinion.                                                    
                                                                                      
By the way, the create_char_line_indexes() function is also used                      
in the next Pull Request related to TAB.